gui/default/syncthing/core/syncthingController.js

```diff
@@ -563,9 +581,50 @@ angular.module('syncthing.core')
                 }
                 $scope.connections = data;
                 console.log("refreshConnections", data);
+
+                refreshRemoteGUI(data);
             }).error($scope.emitHTTPError);
         }
 
+        function refreshRemoteGUI(connections) {
+            if (!$scope.showRemoteGUI) {
+                $scope.idToRemoteGUI = {}
+                return
+            }
+            var newCache = {};
+            for (var id in connections) {
+                if (!(id in $scope.devices)) {
+                    // Avoid errors when called before first updateLocalConfig()
+                    continue;
+                }
+                var port = $scope.devices[id].remoteGUIPort;
+                if (port <= 0
+                    || !connections[id].address
+                    || connections[id].type.includes("relay")) {
+                    // Relay connections never work as desired here, nor incomplete addresses
+                    $scope.idToRemoteGUI[id] = "";
+                    continue;
+                }
+                var newAddress = "http://" + replaceAddressPort(connections[id].address, port);
```

Comment:
This doesn't work when the GUI is loaded over HTTPS, as the request is blocked by the browser.

<img width="862" alt="Screen Shot 2020-11-10 at 12 14 58" src="https://user-images.githubusercontent.com/125426/98667252-7575ae80-234e-11eb-8b24-4a703d4ad786.png">

Reply:
Should we just handle the error?

We can set upgrade insecure url in the options request header. Is this acceptable?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Upgrade-Insecure-Requests

Reply:
I will see if it solves it tomorrow.

Reply:
This seems to be a Really Hard Problem.

Content-Security-Policy basically always disallows mixed content (HTTP request from a script loaded through HTTPS). There are some [additional headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/connect-src) to control that, but apparently can [only further restrict access](https://stackoverflow.com/questions/37043719/how-can-i-allow-mixed-contents-http-with-https-using-content-security-policy-m).

On the other hand, doing probe requests to HTTPS directly will fail almost always. Either the remote GUI uses the auto-generated self-signed TLS certificate, or it has a proper TLS cert, which does not match the IP address we're using to contruct the remote GUI address. Either case leads to the browser blocking the request because of an insecure connection.

One solution on the browser side would be to bypass CSP for OPTIONS requests completely, since these provide no displayed content and therefore pose no security threat. Still a possible privacy problem though.

What we could do on the Syncthing side is looking at the argument of the request error callback. If the request was blocked by the browser already, that `err` argument is `null`. Generating an `http://` link in that case still works because clicking on that is not blocked by CSP (as tested under Firefox). Such a check produces some false-positives, though, as there are many reasons why the browser could have blocked. The OPTIONS method not being allowed in the preflight request (on versions before #7079 was implemented) exhibits the same `err === null` condition for example.

Reply:
Perhaps we should remove the probing in its entirety and just generate an url with the hope of it being useful. I very much wanted the probing approach to work but it seems to be low value...

Reply:
Indeed, ironically browsers have become a rather difficult environment for network programming.  I'm also quite disappointed by all these restrictions, but they do make sense in the name of security within a WWW browsing world.

Especially the problem with link-local IPv6 addresses being unsupported is very unfortunate.  They are very common as a Syncthing connection on LANs, but often unpredictable. If using Bonjour / mDNS or similar, browsers will happily make such connections, but not with IPv6 literals :-( That's still a problem even without trying to probe, but maybe more bearable.